### ISSUE TITLE
Would this help me secure a REST API with SAML?

### ISSUE THREAD

**Issue by cesdperez** [ulisesbocchio/spring-boot-security-saml#34]

**Comment by samtheprogrammer** [comment]
yes it can i had it work with netflix zuul as a service gateway. So every thing below the gateway was protected which allowed me to not have any security logic in my service code

**Comment by ulisesbocchio** [comment]
that's great @samtheprogrammer thanks for jumping in. Would you be able to produce a sample app that we can add to the examples repo?

**Comment by samtheprogrammer** [comment]
Sure i can do that

**Comment by ulisesbocchio** [comment]
awesome, thanks!

**Comment by cesdperez** [comment]
@samtheprogrammer that sounds interesting. I'd totally appreciate a sample too!

**Comment by ulisesbocchio** [comment]
hey @samtheprogrammer any progress on the sample app?

**Comment by aenrosello** [comment]
Any update on the sample? @samtheprogrammer

**Comment by samtheprogrammer** [comment]
sorry been busy with work. Yeah i have the sample

**Comment by freedonaab** [comment]
I would also be very interessted by an example of how to solve this problem. I'll do some research on my side in the meantime and post it here if I find anything.

**Comment by tomiok** [comment]
@ulisesbocchio Hi Ulises and team, is the sample ready for check out?

**Comment by samtheprogrammer** [comment]
I added the code here https://github.com/samtheprogrammer/springboot-saml-gateway
you guys can check it out

**Comment by java-dude** [comment]
> I added the code here https://github.com/samtheprogrammer/springboot-saml-gateway
> you guys can check it out

@samtheprogrammer Thanks for the demo which I can download and run. Is there a username and password to log in to the demo?

**Comment by samtheprogrammer** [comment]
@java-dude  i think it was user: dough1234321@gmail.com pass: Test1234!

**Comment by java-dude** [comment]
> @java-dude i think it was user: [dough1234321@gmail.com](mailto:dough1234321@gmail.com) pass: Test1234!

@samtheprogrammer Thanks, that worked! Nice when a simple demo helps to explain some tricky concepts :)

@samtheprogrammer @ulisesbocchio and team
Since you originally put this together it seems Spring security team completely overhauled the SAML core library back in Oct 2019 with a major v2 release so would you recommend doing this a different way now? Again I'm looking for a simple example being new to SAML.

We  do have a Zuul gateway in front of all our microservices which currently implements spring security access via JWT and is called by our webapps when they want to access the microservices. But now we have a separate requirement from the business to implement SAML authentication for our web apps (not behind gateway) and I'm wondering whether to implement it in the same gateway or if possible as a separate service called by the webapps via the gateway.

Any pointers would be appreciated from experienced people such as yourselves who have already travelled down this path :)

**Comment by samtheprogrammer** [comment]
@java-dude  what you asking is more of a design question. I would prefer same gateway for all apps and microservices, pass user context in headers or some other form to the microservices. This way you have single central piece doing authentication this doesn't mean you cant scale it you can always run multiple instances of you gateway behind a load balancer. It will be easier to maintain if you decide to change authentication protocol in the future. Also lot of load balancers support saml like nginx if you are looking for more performance

**Comment by ulisesbocchio** [comment]
I haven't tried the new SAML v2 library from Spring. Last time I checked it wasn't fully fleshed out. It looks to me that's still the case. I agree with you that it's better to keep the authentication implementation on your gateway. The one aspect you'd have to deal with is that SAML is not an API authentication protocol, so you can't really use SAML after the user signs in for future API requests. The are usually 2 options around this:
1 - Use session. Once your user is authenticated through SAML your Spring app can generate an authenticated session and push a session cookie to the client. Then session is then retrieved on each request to remember the user. You'd need a central session store (Redis, Database) and the Spring session plugin. This is not the most secure approach so you'd have to keep the sessions relatively short lived, forcing the user to re-authenticate every so often.
2 - Use a different protocol. OAuth2 is typically used for this purpose. One option you have is to use Spring CF UAA (https://starkandwayne.com/blog/configure-uaa-with-saml-as-service-provider/) with a SAML Service Provider. Your apps only speak OAuth2 and UAA will do the heavy lifting and provide an OAuth2 Token after you successfully login through SAML. Or you can create create something similar to UAA yourself, having a Spring Boot app with 2 inter-connected security realms, one SAML, one OAuth2 and plugin the SAML auth flow to the beginning of your OAuth flow. Something like OAuth2 -> SAML -> OAuth2. Which is essentially what UAA does

**Comment by java-dude** [comment]
> @java-dude what you asking is more of a design question. I would prefer same gateway for all apps and microservices, pass user context in headers or some other form to the microservices. This way you have single central piece doing authentication this doesn't mean you cant scale it you can always run multiple instances of you gateway behind a load balancer. It will be easier to maintain if you decide to change authentication protocol in the future. Also lot of load balancers support saml like nginx if you are looking for more performance

@samtheprogrammer  Thanks for the feedback. Yes we do have multiple instances behind a load balancer but the webapps are  external developed by another team and thats one of the classic uses of an API gateway I have seen .... to put the gateway inbetween external web clients and your microservices ...  although having SAML authentication in your gateway before hitting web clients  makes life easier hence the dilemma. Not looked at nginx support for SAML sounds an interesting option.

**Comment by java-dude** [comment]
> I haven't tried the new SAML v2 library from Spring. Last time I checked it wasn't fully fleshed out. It looks to me that's still the case. I agree with you that it's better to keep the authentication implementation on your gateway. The one aspect you'd have to deal with is that SAML is not an API authentication protocol, so you can't really use SAML after the user signs in for future API requests. The are usually 2 options around this:
> 1 - Use session. Once your user is authenticated through SAML your Spring app can generate an authenticated session and push a session cookie to the client. Then session is then retrieved on each request to remember the user. You'd need a central session store (Redis, Database) and the Spring session plugin. This is not the most secure approach so you'd have to keep the sessions relatively short lived, forcing the user to re-authenticate every so often.
> 2 - Use a different protocol. OAuth2 is typically used for this purpose. One option you have is to use Spring CF UAA (https://starkandwayne.com/blog/configure-uaa-with-saml-as-service-provider/) with a SAML Service Provider. Your apps only speak OAuth2 and UAA will do the heavy lifting and provide an OAuth2 Token after you successfully login through SAML. Or you can create create something similar to UAA yourself, having a Spring Boot app with 2 inter-connected security realms, one SAML, one OAuth2 and plugin the SAML auth flow to the beginning of your OAuth flow. Something like OAuth2 -> SAML -> OAuth2. Which is essentially what UAA does

@ulisesbocchio Thanks for the detailed feedback. Hmm so as i suspected seems like v2 library is in flux and going for Spring saml v1 library may still be the way to go. So was originally thinking to try keep the current Spring Security JWT authentication we already have in the gateway for APIs but also implementing the SAML user authentication for the webapps for SSO here somehow or as a service called by the gateway although not sure how this would work in reality as being new to this. Will look at both options you suggested above though - thanks for the link.

**Comment by kapsebharat1** [comment]
> @java-dude i think it was user: [dough1234321@gmail.com](mailto:dough1234321@gmail.com) pass: Test1234!

@samtheprogrammer Looks like its changed, can you share updated one
